### ISSUE TITLE
Fine graine the OVERRIDE_PV_CONFIG_WITH_IMAGE_CONFIG process

### ISSUE THREAD

**Issue by arnaud-deprez** [openshift/jenkins#476]
See comments [here](https://github.com/openshift/jenkins/commit/84c2a3c71200b9626e83d2a39e095f75b615c5ed#commitcomment-26745639).

Basically when you use this option OVERRIDE_PV_CONFIG_WITH_IMAGE_CONFIG, every file and directory in the git repository replaces the one in the volume.
For the directory, it may be too "hard" and so it may be more reasonable to keep the directory in the volume and only replace the files in it.

As I did that code in the first place, I'll handle it.

**Comment by arnaud-deprez** [comment]
@jr00n here is why I didn't did that in the first place: https://github.com/openshift/jenkins/pull/420#discussion_r151268350

**Comment by bparees** [comment]
I appreciate the use case but i think we're going down a deep rathole of config management here.

I don't want to start adding a dozen flags to let people choose exactly what parts of the image config they want vs what they want to keep from the PV.

At the end of the day you can solve this by:

1) building a new image w/ the config you want
2) building a new image w/ your own "Run" script that implements whatever image config copying behavior you want

So i'm currently against baking this complexity into our run scripts.

/cc @gabemontero

**Comment by arnaud-deprez** [comment]
@bparees I agree, we can't handle each use case.

FYI, my idea wasn't to add a flag or so but "simply" going recursively to each file in subfolder. 
*Advantage:*
* This would at keep existing content in directories and override just the needed files
*Inconvenients:*
* But if for whatever reason, we decide to rename a file maintain in git (like a job), this configuration will remain in the volume and so you will have to delete it manually.
* It's a bit more complex to maintain and we will probably receive other complains about that because of this first point.

Basically, as I said in #420, as a best practices, once you start configuring jenkins in your image, you should do it for every single configuration so your infrastructure is managed as code and fully reproducible.

**Comment by gabemontero** [comment]
I think we have consensus.  At least enough of it.  Closing out.

**Comment by jr00n** [comment]
@arnaud-deprez Can you help me explain what to do in my case?

I work at a company with a lot of java development teams. They al going to use jenkins in there openshift projects.

On my department we provide Software Development Tools like Jenkins etc. 

We make a specific Jenkins image (based on the openshift/jenkins s2i image) with specific company config and plugins and some default jobs (as example) so they can use my image.

After a while i change the image because something is change in our environment or way of using Jenkins so we push a new image for the teams.

All the Jenkins customers are going to use that new image, and we like them to use our config, plugins etc and maybe a new demo job.

But meanwhile they have allready jobs their own...so in the current situatie they get lost from the PV.

In what situation would you like override everything like you've made? 

What would you do in my case?

**Comment by bparees** [comment]
@jr00n you should provide a customized jenkins image (inheriting from ours) which does the following:

1) sets its entrypoint/cmd to a new startup script provided by you
2) that new startup script can implement any config copying logic you want
3) that new startup script should then invoke the original jenkins startup script from the image (w/o the override env obviously), passing all args that came in.

**Comment by arnaud-deprez** [comment]
@jr00n you are in a mix-in situation where some jobs are defined in the image and are immutables and some are defined by the customers outside the image.

This is generally not recommended as docker container should be immutables. 
You can find a lot of blog/article on the web that explain it but here is one from RedHat https://developers.redhat.com/blog/2016/02/24/10-things-to-avoid-in-docker-containers/.

If you follow this guideline, this will help you to answer these questions: 
* What happens if your jenkins instance die and you just need to restart it ? 
* What happens if your persistent volume gets corrupted and you need to clean it up ? 

So ideally, I would recommend that your customer extends your image and define new jobs.
To facilitate this, I recommend you to use [Jenkins Groovy initialization scripts](https://wiki.jenkins.io/display/JENKINS/Groovy+Hook+Script) + [Groovy Job DSL plugin](https://jenkinsci.github.io/job-dsl-plugin/)

Then if it's not possible, you can do what @bparees just proposed which is the most straightforward solution I think.
